### PR TITLE
core: layout change - introduce prev field in pmemfile_block

### PR DESCRIPTION
**Pull request by GBuella**
This change is [Reviewable](https://reviewable.io/reviews/marcinslusarz/pmemfile/6)

### CODE REVIEW

File: tests/core/rw/rw.cpp [hunk 1]
```diff
@@ -334,10 +334,10 @@ TEST_F(rw, 2)
 	    {}}));
 
 	if (env_block_size == 4096)
-		EXPECT_TRUE(test_pmemfile_stats_match(pfp, 2, 0, 609, 0,
+		EXPECT_TRUE(test_pmemfile_stats_match(pfp, 2, 0, 0x32c, 0,
```
Comment (marcinslusarz) [inline tests/core/rw/rw.cpp]: Why in hex?

File: src/libpmemfile-core/data.c [hunk 2]
```diff
@@ -365,9 +381,9 @@ file_allocate_range(PMEMfilepool *pfp, struct pmemfile_vinode *vinode,
 
 			block->offset = offset;
 
-			block->next =
-			    (TOID(struct pmemfile_block))
-			    pmemobj_oid(vinode->first_block);
+			block->next = blockp_as_oid(vinode->first_block);
+			block->prev = TOID_NULL(struct pmemfile_block);
+			vinode->first_block->prev = blockp_as_oid(block);
```
Comment (marcinslusarz) [inline src/libpmemfile-core/data.c]: Don't we need to snapshot it first?

File: src/libpmemfile-core/data.c [hunk 3]
```diff
@@ -412,11 +429,12 @@ file_allocate_range(PMEMfilepool *pfp, struct pmemfile_vinode *vinode,
...
-			previous->next =
-			    (TOID(struct pmemfile_block))pmemobj_oid(block);
+			previous->next = blockp_as_oid(block);
+
+			block->prev = blockp_as_oid(previous);
+			block->next = blockp_as_oid(next);
 
-			block->next =
-			    (TOID(struct pmemfile_block))pmemobj_oid(next);
+			next->prev = blockp_as_oid(block);
```
Comment (marcinslusarz) [inline src/libpmemfile-core/data.c]: Again, snapshot?

File: tests/core/rw/rw.cpp (hunk 1 above)
Comment (GBuella) [inline tests/core/rw/rw.cpp]: When the test failed, it was printed in hex, something like
0x261 != 0x32c
I did not see a reason to convert it to decimal. Now it is in the same format, as a failed test's output would be.

File: src/libpmemfile-core/data.c (hunk 2 above)
Comment (GBuella) [inline src/libpmemfile-core/data.c]: Forgot about it.

File: src/libpmemfile-core/data.c (hunk 3 above)
Comment (GBuella) [inline src/libpmemfile-core/data.c]: Forgot about it.

File: src/libpmemfile-core/data.c (hunk 2 above)
Comment (marcinslusarz) [inline src/libpmemfile-core/data.c]: Then it means this code path is not exercised under pmemcheck. Please write a test case for this code path.

**Comment by marcinslusarz** [comment]
lgtm

**Comment by marcinslusarz** [comment]
Reviewed 1 of 3 files at r1, 1 of 1 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/6)*

**Pull request merged**